### PR TITLE
feat(events): rsvp enrollment end-to-end

### LINKED ISSUE #237: feat(events): participation module MVP
## Problem Statement

A He4rt Developers organiza meetups presenciais, workshops e conferências (incluindo um meetup em breve e uma conferência de 2 dias no final do ano), mas não possui um sistema próprio para gerenciar o ciclo de participação dos membros. Hoje não há como:
- Controlar inscrições com capacidade limitada e lista de espera
- Verificar presença de quem realmente compareceu (check-in)
- Recompensar participação com XP no sistema de gamificação existente
- Aprovar candidatos a eventos com vagas limitadas (workshops)

## Solution

Construir o módulo de participação em eventos (`app-modules/events/`) com lifecycle completo: enrollment → check-in → attendance → XP reward. O sistema suporta 3 métodos de entrada (RSVP, RSVP+check-in, Application com formulário dinâmico), 3 métodos de check-in (manual, código numérico, QR code), waitlist com promoção FIFO, e integração com o módulo de gamificação via domain events.

## User Stories

1. As a participant, I want to RSVP to a meetup with one click, so that I can confirm my presence quickly
2. As a participant, I want to join a waitlist when an event is full, so that I can be promoted if someone cancels
3. As a participant, I want to be automatically promoted from the waitlist when a slot opens, so that I don't miss the opportunity
4. As a participant, I want to cancel my enrollment before the deadline, so that my slot is freed for others
5. As a participant, I want to check in at a meetup by typing a numeric code, so that my presence is verified
6. As a participant, I want to check in at a conference by showing my QR code badge, so that attendance is tracked per day
7. As a participant, I want to receive XP when I confirm, check in, and complete attendance, so that my participation is rewarded in the gamification system
8. As a participant, I want to fill an application form to apply for a workshop, so that the organizer can evaluate my candidacy
9. As a participant, I want to be notified when my application is approved or rejected, so that I know my status
10. As a participant, I want to check in on each day of a multi-day conference, so that my attendance per day is tracked
11. As a participant, I want to use the same QR code across all event days, so that I only need one badge
12. As a participant, I want to check in via Discord/Twitch bot by typing a command with the code, so that I can confirm presence in online events
13. As a participant, I want to see my enrollment status for each event, so that I know where I stand
14. As an organizer, I want to create events with configurable enrollment policies, so that each event has appropriate participation rules
15. As an organizer, I want to set capacity limits with optional waitlist, so that I can control attendance
16. As an organizer, I want to choose the enrollment method (RSVP, RSVP+check-in, Application), so that the entry flow matches the event type
17. As an organizer, I want to choose the check-in method (manual, numeric code, QR code), so that presence verification fits the event format
18. As an organizer, I want to define a dynamic application form (JSONB schema), so that I can collect relevant information per event
19. As an organizer, I want to approve or reject applications with a reason, so that participants understand the decision
20. As an organizer, I want to generate time-limited numeric codes bound to a specific date, so that check-in is controlled per day
21. As an organizer, I want to manually check in participants, so that I can handle edge cases (late arrivals, code issues)
22. As an organizer, I want to override a participant's status (e.g., no_show → attended), so that I can correct errors
23. As an organizer, I want to set a cancellation deadline in hours before the event, so that I have predictability on attendance
24. As an organizer, I want to configure XP rewards per event (RSVP, check-in, attendance), so that different events reward differently
25. As an organizer, I want to set attendance_requirement (all_days, any_day, minimum_days), so that multi-day events have clear completion criteria
26. As an organizer, I want to see all enrollments with their current status, so that I can manage the event
27. As an organizer, I want a full audit trail of every status transition, so that I can investigate disputes
28. As the system, I want to automatically mark checked_in → attended after the event ends, so that XP is awarded without manual intervention
29. As the system, I want to automatically mark confirmed → no_show after the event ends, so that non-attendees are identified
30. As the system, I want enrollment capacity checks to be atomic (lockForUpdate), so that overbooking is impossible under concurrency

### Architecture

- **Module**: `app-modules/events/` following existing modular architecture (`internachi/modular`)
- **Namespace**: `He4rt\Events\*`
- **Panels**: Admin panel = organizer, App panel = participant
- **Table prefix**: All tables prefixed with `events_` for clear identification in database

### State Machine

- Implemented as a PHP backed enum (`EnrollmentStatusEnum`) with `canTransitionTo()` method
- **NOT** using `spatie/laravel-model-states` — transitions validated in Action classes explicitly
- 8 states: `pending`, `confirmed`, `waitlisted`, `checked_in`, `attended`, `cancelled`, `rejected`, `no_show`
- Terminal states: `attended` (success), `cancelled`, `rejected`, `no_show`
- See ADR-0001

### Schema (7 tables)

**`events`** — Core event record:
- `id`, `tenant_id`, `event_type` (meetup/workshop/conference), `active`, `slug`, `title`, `description`, `location`, `starts_at`, `ends_at`, `timestamps`
- No denormalized counters — derived by query

**`events_enrollment_policies`** — 1:1 configuration per event:
- `event_id` (unique), `enrollment_method`, `check_in_method`, `capacity` (nullable = unlimited), `has_waitlist`, `attendance_requirement` (all_days/any_day/minimum_days), `cancellation_deadline_hours`, `xp_reward_rsvp`, `xp_reward_checkin`, `xp_reward_attendance`, `application_form_schema` (JSONB), `timestamps`

**`events_enrollments`** — Hub central (1 per user+event):
- `id`, `event_id`, `user_id`, `status` (enum), `waitlist_position` (nullable), `application_data` (JSONB), `rejection_reason`, `enrolled_at`, `confirmed_at`, `checked_in_at`, `attended_at`, `cancelled_at`, `is_public`, `timestamps`
- UNIQUE constraint on `(event_id, user_id)`

**`events_enrollment_transitions`** — Append-only audit trail:
- `id`, `enrollment_id`, `from_status`, `to_status`, `triggered_by` (system/user/admin), `actor_user_id` (nullable), `reason` (nullable), `created_at`
- Written explicitly by Actions, NOT by database triggers (ADR-0003)

**`events_check_ins`** — 1:N per enrollment (one per day):
- `id`, `enrollment_id`, `method` (manual/numeric_code/qr_code), `payload` (JSONB), `event_date`, `checked_in_at`, `timestamps`
- Multi-day events: one check-in record per day attended (ADR-0004)
- Event days derived from `events.starts_at`/`ends_at` date range — no separate days table

**`events_check_in_codes`** — Organizer-generated numeric codes:
- `id`, `event_id`, `code`, `event_date`, `valid_from`, `valid_until`, `max_uses` (nullable), `uses_count`, `timestamps`
- Bound to specific date — one code per day

**`events_qr_tokens`** — One per enrollment, reusable across days:
- `id`, `enrollment_id`, `token` (unique), `expires_at` (nullable), `timestamps`

### XP Integration

- Events module publishes domain events: `EnrollmentConfirmed`, `ParticipantCheckedIn`, `ParticipantAttended`
- Gamification module listens and calls `IncrementExperience` on Character
- XP amounts come from `events_enrollment_policies` fields, passed as payload in domain events
- No XP ledger in Events module — auditability via `events_enrollment_transitions`
- Idempotency responsibility on Gamification listener
- See ADR-0002

### Bot Integration

- Discord/Twitch bot dispatches a domain event (e.g., `CheckInRequested`) with user identifier and code
- Events module has a listener that validates the code and processes the check-in
- Bot is pure transport — all business logic in Events module
- Bot must listen for response event (`CheckInProcessed`) to reply to user in chat
- Consistent with existing Bot Discord → Moderation pattern in CONTEXT-MAP
- See ADR-0005

### Enrollment Flow — Concurrency

- Capacity check uses `DB::transaction` with `lockForUpdate()` on the enrollment policy
- Count confirmed/checked_in/attended enrollments → compare against capacity → insert or waitlist
- Waitlist promotion on cancellation: FIFO (ordered by `enrolled_at`), atomic update + notification dispatch

### Scheduled Job

- Runs after `event.ends_at` passes
- Batch transitions: `checked_in → attended` (if attendance_requirement met), `confirmed → no_show`
- Dispatches XP domain events for newly `attended` enrollments
- Single job handles both transitions per event

### Application Form

- Dynamic form schema stored as JSONB in `events_enrollment_policies.application_form_schema`
- Filament form builder for organizer to define questions
- Participant responses stored in `events_enrollments.application_data` as JSONB
- Organizer reviews in Admin panel with approve/reject actions

### Event Types

```php
enum EventTypeEnum: string {
    case Meetup = 'meetup';
    case Workshop = 'workshop';
    case Conference = 'conference';
}
```

## Testing Decisions

- **Good tests**: test external behavior through Actions and domain events, not internal implementation. Assert state transitions, database state, and dispatched events.
- **Test location**: `app-modules/events/tests/Feature/` and `app-modules/events/tests/Unit/`
- **Prior art**: existing tests in `app-modules/moderation/tests/` and `app-modules/bot-discord/tests/`
- **What to test**:
  - Each Action (EnrollUser, CancelEnrollment, CheckIn, ApproveApplication, RejectApplication, PromoteFromWaitlist) — happy path + edge cases
  - State machine transitions — valid and invalid
  - Concurrency: capacity enforcement under race conditions
  - Waitlist promotion FIFO ordering
  - Multi-day check-in with attendance_requirement validation
  - Scheduled job (attended + no_show marking)
  - Domain event dispatch (assert events are fired, don't test Gamification's listener here)
  - Cancellation deadline enforcement
  - Numeric code validation (expiry, max_uses, wrong date)
  - QR token validation (expiry, reuse across days)
- **Factories**: create factories for EventModel, EventEnrollmentPolicy, EventEnrollment, EventCheckIn, EventCheckInCode, EventQrToken
- **Use Pest** with `php artisan make:test --pest`
- **ActingAs**: all Filament panel tests require `$this->actingAs(User::factory()->create())`

## Out of Scope

- Networking between participants (profiles, connections, discovery)
- Referral / invite links with tracking
- Magic link check-in method
- Geolocation check-in method
- Sponsors association with events
- Timeline / activity feed display
- Call for Papers / Submissions (CFP)
- Agenda / schedule display
- Paid events / payment integration
- No-show penalties affecting future eligibility (data tracked, consequences deferred)
- Event days as explicit table (derived from date range)
- Denormalized counters on events table

## Further Notes

- The events module was previously gutted (all tables dropped in migration `2026_05_16_195205`). This PRD rebuilds the module from scratch with the new participation-focused architecture.
- The previous module had: events, events_talks, events_attendees, sponsors, events_sponsors, events_agenda, event_submission_speakers. The new module replaces this with a fundamentally different schema focused on enrollment lifecycle.
- Existing ADRs document key architectural decisions: see `app-modules/events/docs/adr/0001-0005`.
- `CONTEXT-MAP.md` at repo root has been updated to include Events and Gamification bounded contexts with dependency diagram.
- `app-modules/events/CONTEXT.md` contains the full domain glossary and state machine reference.
- Future phases (post-MVP): magic_link/geolocation check-in, CFP/submissions, agenda display, networking, referral, payment integration, no-show penalties.

### LINKED ISSUE #240: feat(events): RSVP enrollment end-to-end
## Parent

#237

## What to build

Implement the RSVP enrollment flow end-to-end: participant clicks "Confirm presence" → system creates enrollment with status `confirmed` → transition is recorded in audit trail → participant sees their enrollment in the App panel.

**Action: `EnrollUserAction`**
- Accepts event + user + enrollment method context
- For `rsvp` and `rsvp_checkin` methods: creates enrollment with status `confirmed` (capacity validation comes in the next slice)
- Wraps in `DB::transaction`
- Records transition in `events_enrollment_transitions` (from_status=null, to_status=confirmed, triggered_by=user)
- Returns the created enrollment

**Domain event (fire-and-forget, for future listeners):**
- Dispatches `EnrollmentConfirmed` implementing `ShouldDispatchAfterCommit`
- Payload: enrollment_id, event_id, user_id, xp_reward_rsvp (from policy)
- No listener in this slice — just the event dispatch

**App panel (participant view):**
- "My Events" page listing user's enrollments with status badges
- Event detail page with "Confirm Presence" button (visible when enrollment_method is rsvp or rsvp_checkin and user is not yet enrolled)
- Button dispatches EnrollUserAction

**Admin panel (organizer view):**
- Enrollments RelationManager on Event resource showing all enrollments with status, enrolled_at
- Read-only for now (management actions come in later slices)

**Validation rules:**
- User cannot enroll twice in same event (UNIQUE constraint)
- Event must be active and not past (`starts_at > now()`)
- Enrollment method must be `rsvp` or `rsvp_checkin` (application flow is a separate slice)

## Acceptance criteria

- [ ] Participant can RSVP to an event from the App panel
- [ ] Enrollment is created with status `confirmed` and `enrolled_at` timestamp
- [ ] Transition record is written to `events_enrollment_transitions`
- [ ] `EnrollmentConfirmed` domain event is dispatched after transaction commit
- [ ] Duplicate enrollment attempt returns appropriate error (409 or validation error)
- [ ] Enrolling in a past event is rejected
- [ ] Admin panel shows enrollments per event
- [ ] Participant sees their enrollments in "My Events" page
- [ ] Feature tests cover: successful enrollment, duplicate prevention, past event rejection
- [ ] Pint passes

## Blocked by

- #239 (Foundation)

### PR DESCRIPTION
**Pull request by BrunaDomingues**
## Summary

Implements the **Events** bounded context (foundation) and the **RSVP enrollment** flow end-to-end: participant confirms presence → enrollment created as `confirmed` or `waitlisted` (when at capacity with waitlist) → transition recorded in audit trail → enrollment visible in App and Admin panels.

### RSVP enrollment

- **`EnrollUserAction`**: validates event (published, upcoming, RSVP/rsvp_checkin), resolves initial status (`confirmed` vs `waitlisted` vs rejected when full), creates enrollment, writes `EnrollmentTransition` (`from_status=null`, `to_status=<initial>`, `triggered_by=user`), dispatches `EnrollmentConfirmed` only when status is `confirmed` — all inside `DB::transaction` with `lockForUpdate` on policy and duplicate handling via `UniqueConstraintViolationException`
- **`EnrollUserDTO`**: input object built from `Event` + `User` (status, dates, policy capacity/waitlist, XP reward)
- **Domain event**: `EnrollmentConfirmed` implements `ShouldDispatchAfterCommit` (payload: `enrollment_id`, `event_id`, `user_id`, `xpRewardOnConfirmed`) — no listener in this slice
- **`EnrollmentException`**: `alreadyEnrolled`, `eventPast`, `eventNotActive`, `invalidEnrollmentMethod`, `eventFull`

### Capacity & waitlist

- When `enrollment_policy.capacity` is set and confirmed count ≥ capacity:
  - **`has_waitlist=true`** → enrollment `waitlisted` with `waitlist_position` (no `EnrollmentConfirmed`)
  - **`has_waitlist=false`** → `EnrollmentException::eventFull()`
- App UI: **Confirm Presence** hidden when full without waitlist; success notification differs for confirmed vs waitlisted; **My Events** shows status badges (including waitlisted)

### App panel (participant)

- Events listing (`EventsPage` / `EventsList`) — published upcoming events
- Event detail (`EventPage` / `EventDetail`) — **Confirm Presence** button
- My Events (`MyEventsPage` / `MyEventsList`) — enrollments with status badges; enrolled users can open detail for `completed` events

### Admin

- **`EventResource`**: form/infolist/table aligned with `status` column (`EventStatus` enum) instead of legacy `active` toggle
- Unique event **slug per tenant** validation on create/edit
- Read-only **Enrollments** relation manager

### Foundation (Events module — same branch vs `4.x`)

Schema, models (`Event`, `Enrollment`, `EnrollmentPolicy`, `EnrollmentTransition`, check-in models), enums, factories, migrations, ADRs, `CONTEXT.md`, admin resource scaffolding, unit tests for enums.

### Architecture
Events module ├── Enrollment domain │ ├── EnrollUserAction │ ├── EnrollUserDTO │ ├── EnrollmentConfirmed (domain event) │ └── EnrollmentException ├── Event domain (scopes: published, upcoming, viewableByParticipant) └── Check-in models (schema only — no flow in this PR)

App panel ├── EventsPage / EventsList ├── EventPage / EventDetail (+ Confirm Presence) └── MyEventsPage / MyEventsList

Admin panel └── EventResource (+ EnrollmentsRelationManager)

### Files (high level)

| Area | Count / notes |
|------|----------------|
| Migrations | 7 (+ drop legacy tables) |
| Actions | `EnrollUserAction` |
| DTOs | `EnrollUserDTO` |
| Domain events | `EnrollmentConfirmed` |
| Exceptions + lang | `EnrollmentException`, en/pt_BR exceptions + pages + enums |
| App panel | 3 Filament pages, 3 Livewire components, 7 Blade views (incl. partial) |
| Admin | `EventResource` schemas + relation manager |
| Tests | `EnrollUserActionTest`, `RsvpEnrollmentTest`, `EventResourceTest`, `EventFactoriesTest`, enum unit tests |
| Docs | 5 ADRs, `CONTEXT.md`, `CONTEXT-MAP.md` |

### Out of scope (future slices)

- Gamification listener for XP (`EnrollmentConfirmed` subscriber)
- Application enrollment flow (approve/reject)
- Enrollment management actions in admin (promote from waitlist, cancel, etc.)
- Check-in / QR flows
- Bot integration

Closes #240  
Parent: #237  
Related foundation PR: #249 (included in this branch when targeting `4.x`)

---

## Test plan

- [ ] `php artisan migrate`
- [ ] `php artisan test --filter=EnrollUserActionTest`
- [ ] `php artisan test --filter=RsvpEnrollmentTest`
- [ ] `php artisan test --filter=EventResourceTest`
- [ ] `php artisan test --filter=EventFactoriesTest`
- [ ] `./vendor/bin/pint --test`

### Admin

- [ ] Create event: **Published**, future `starts_at`, enrollment method **RSVP**
- [ ] Set `capacity` and toggle `has_waitlist` on enrollment policy
- [ ] Edit event → **Enrollments** tab shows participants

### App — happy path

- [ ] `/app/{tenant}/events` — event appears in list
- [ ] Open event → **Confirm Presence** → success notification
- [ ] `/app/{tenant}/my-events` — enrollment shows **Confirmed** badge

### App — capacity / waitlist

- [ ] Fill event to capacity **without** waitlist → **Confirm Presence** hidden; direct enroll returns `event_full`
- [ ] Fill event to capacity **with** waitlist → button still visible; enroll creates **Waitlisted** badge and waitlist notification

### App — edge cases

- [ ] Duplicate RSVP → error notification (`already_enrolled`)
- [ ] Past event → no **Confirm Presence** button
- [ ] Draft / non-RSVP event → not enrollable via RSVP
- [ ] Enrolled user on **completed** event → detail page still accessible

### CODE REVIEW

File: app-modules/panel-app/src/Livewire/Events/EventDetail.php [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+        $method = $this->event->enrollmentPolicy?->enrollment_method;
+
+        if (!in_array($method, [EnrollmentMethod::Rsvp, EnrollmentMethod::RsvpCheckin], strict: true)) {
+            return false;
+        }
+
+        return $this->event->starts_at->isFuture();
+    }
+
```
Comment (davicbtoliveira) [inline app-modules/panel-app/src/Livewire/Events/EventDetail.php:86]: **canConfirmPresence() não reflete o estado de capacidade do evento**

O método `canConfirmPresence()` verifica apenas duplicidade, método de
inscrição e se o evento ainda não começou. Ele não considera se o evento
está lotado ou se o usuário deveria ser colocado em fila de espera.

Isso pode resultar em uma ação de "confirmar presença" sendo exibida na
UI mesmo quando o domínio deveria bloquear ou enfileirar a inscrição.

File: app-modules/panel-app/src/Livewire/Events/MyEventsList.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    public function getEnrollmentsProperty(): Collection
+    {
+        return Enrollment::query()
+            ->with(['event'])
+            ->where('user_id', auth()->id())
+            ->whereHas('event', fn (Builder $query) => $query->where('tenant_id', filament()->getTenant()->getKey()))
+            ->latest('enrolled_at')
+            ->get();
+    }
+
```
Comment (davicbtoliveira) [inline app-modules/panel-app/src/Livewire/Events/MyEventsList.php:27]: **Listagem pode exibir links que resultam em 404**

A listagem exibe todas as inscrições do usuário sem filtrar
pelo status do evento. Já a página de detalhe do evento só permite acesso
a eventos com status Published.

Se um evento inscrito for movido para Completed, Cancelled ou tiver sua
publicação removida, o link na listagem continuará sendo renderizado mas
resultará em 404 ao ser acessado, quebrando o histórico de navegação do
participante.

**Sugestão:** Alinhar o filtro da listagem com as mesmas regras de acesso
da página de detalhe, ou adaptar a página de detalhe para exibir eventos
nos demais estados de forma adequada (ex: modo somente leitura para
eventos concluídos/cancelados).

File: app-modules/events/src/Enrollment/Actions/EnrollUserAction.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+        return DB::transaction(function () use ($dto): Enrollment {
+            throw_if(
+                Enrollment::query()
+                    ->where('event_id', $dto->eventId)
+                    ->where('user_id', $dto->userId)
+                    ->exists(),
+                EnrollmentException::alreadyEnrolled(),
+            );
+
```
Comment (davicbtoliveira) [inline app-modules/events/src/Enrollment/Actions/EnrollUserAction.php]: **[Bug] Inscrição ignora capacidade e fila de espera**

O `EnrollUserAction` sempre define `EnrollmentStatus::Confirmed`
independentemente da capacidade atual do evento ou da flag `has_waitlist`.
Isso conflita diretamente com os estados de Events, que espera
que eventos lotados coloquem novos inscritos em waitlist.

Como consequência, eventos podem ser sobrecarregados além da capacidade e
o campo `has_waitlist` perde o efeito prático.

File: app-modules/events/src/Enrollment/Events/EnrollmentConfirmed.php [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+final readonly class EnrollmentConfirmed implements ShouldDispatchAfterCommit
+{
+    use Dispatchable;
+
+    public function __construct(
+        public string $enrollmentId,
+        public string $eventId,
+        public string $userId,
+        public int $xpRewardRsvp,
```
Comment (GabrielFVDev) [inline app-modules/events/src/Enrollment/Events/EnrollmentConfirmed.php]: aq como 'xpRewardRsvp', mas no dto dele ta como 'xpRewardOnConfirmed'.
Num seria melhor deixar tudo com o msm nome?

File: app-modules/events/lang/pt_BR/exceptions.php [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+return [
+    'already_enrolled' => 'Você já está inscrito neste evento.',
+    'event_past' => 'Este evento já começou e não aceita mais inscrições.',
+    'event_not_active' => 'Este evento não está disponível para inscrição.',
+    'invalid_enrollment_method' => 'Este evento não aceita inscrições via RSVP.',
```
Comment (GabrielFVDev) [inline app-modules/events/lang/pt_BR/exceptions.php]: acho q o texto seria legal algo como "Esse evento exige inscrição via formulario, não RSVP", ou eu to delirando?

**Review by GabrielFVDev — COMMENTED** [review]
De geral tá bão, so vou ajustar o teste de migration, que acabou dando erro:

  2026_04_18_000001_add_unique_index_to_messages_provider_message_id ................................................................... 2.73ms DONE
  2026_04_19_000001_change_external_identities_metadata_to_jsonb ....................................................................... 1.14ms FAIL


Novamente, de geral ta bão

File: app-modules/events/src/Enrollment/Actions/EnrollUserAction.php [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    public function handle(EnrollUserDTO $dto): Enrollment
+    {
+        $this->validate($dto);
+
+        return DB::transaction(function () use ($dto): Enrollment {
+            throw_if(
+                Enrollment::query()
+                    ->where('event_id', $dto->eventId)
+                    ->where('user_id', $dto->userId)
+                    ->exists(),
```
Comment (GabrielFVDev) [inline app-modules/events/src/Enrollment/Actions/EnrollUserAction.php]: O cancelamento de RSVP é final por decisao mesmo?, ou queremos permitir re-inscrição enquanto o evento ainda está aberto? 

Parece que o exists() aqui acaba bloqueando o usuário de se inscrever de novo se ele já tem qualquer enrollment pro evento, inclusive Cancelled/Rejected/NoShow. E isso bate com o resto do módulo (esses status são terminais no enum  e a migration tem UNIQUE(event_id, user_id)).

 Resumindo: "cancelou, acabou", não tem como voltar atrás nem se o evento ainda tem vagas.

**Review by GabrielFVDev — APPROVED** [review]
o restante aparenta estar de boa, só dar uma revisada no comentario que deixei.


LGTM

**Comment by hefeus** [comment]
LGTM

File: app-modules/events/src/Enrollment/Actions/EnrollUserAction.php (hunk 6 above)
Comment (BrunaDomingues) [reply app-modules/events/src/Enrollment/Actions/EnrollUserAction.php]: Vai ser feito algo na issue #242

**Review by davicbtoliveira — APPROVED** [review]
LGTM

File: app-modules/events/src/Enrollment/Actions/EnrollUserAction.php [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+                $enrollment->status = $initial['status'];
+                $enrollment->save();
+
+                EnrollmentTransition::query()->create([
+                    'enrollment_id' => $enrollment->id,
+                    'from_status' => null,
+                    'to_status' => $initial['status'],
+                    'actor_id' => $dto->userId,
+                    'triggered_by' => TriggeredBy::User,
+                ]);
```
Comment (danielhe4rt) [inline app-modules/events/src/Enrollment/Actions/EnrollUserAction.php]: Por quê usar os dois padrões? 

Eu prefiro muito mais o `Model::query()->create()`.

File: app-modules/panel-app/src/Livewire/Events/EventDetail.php [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        try {
+            $enrollment = resolve(EnrollUserAction::class)->handle(
+                EnrollUserDTO::fromModels($this->event, $user),
+            );
+
+            unset($this->enrollment, $this->canConfirmPresence);
+
+            $successMessage = $enrollment->status === EnrollmentStatus::Confirmed
+                ? __('events::pages.confirm_presence_success')
+                : __('events::pages.waitlist_success');
```
Comment (danielhe4rt) [inline app-modules/panel-app/src/Livewire/Events/EventDetail.php]: Dentro do enum, faça uma função de helper pra getResponseMessage() e funções para: isConfirmed() e is(outros-enum-cases). Isso facilita a leitura do código.

File: app-modules/events/src/Enrollment/DTOs/EnrollUserDTO.php [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    public function __construct(
+        public string $eventId,
+        public string $userId,
+        public EventStatus $eventStatus,
+        public CarbonInterface $eventStartsAt,
+        public ?EnrollmentMethod $enrollmentMethod,
+        public ?int $capacity,
+        public bool $hasWaitlist,
+        public int $xpRewardOnConfirmed,
+    ) {}
```
Comment (danielhe4rt) [inline app-modules/events/src/Enrollment/DTOs/EnrollUserDTO.php:15]: Pra evitar possíveis race conditions, passe apenas as chaves e resolva todo o resto dentro da action. 

```suggestion
    public function __construct(
        public string $eventId,
        public string $userId
    ) {}
```

File: app-modules/events/src/Enrollment/Actions/EnrollUserAction.php [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+                    ->where('event_id', $dto->eventId)
+                    ->where('user_id', $dto->userId)
+                    ->exists(),
+                EnrollmentException::alreadyEnrolled(),
+            );
+
+            EnrollmentPolicy::query()
+                ->where('event_id', $dto->eventId)
+                ->lockForUpdate()
+                ->first();
```
Comment (danielhe4rt) [inline app-modules/events/src/Enrollment/Actions/EnrollUserAction.php]: Isso dá um lock, mas onde vem o release? E outra: a query é pra travar apenas ou pra reutilizar? Seria bom deixar um pouco mais de contexto aqui.

File: app-modules/events/src/Enrollment/Actions/EnrollUserAction.php [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+        $this->validate($dto);
+
+        return DB::transaction(function () use ($dto): Enrollment {
+            throw_if(
+                Enrollment::query()
+                    ->where('event_id', $dto->eventId)
+                    ->where('user_id', $dto->userId)
+                    ->exists(),
+                EnrollmentException::alreadyEnrolled(),
+            );
```
Comment (danielhe4rt) [inline app-modules/events/src/Enrollment/Actions/EnrollUserAction.php]: Essa linha está duplicada. A mesma validação está no método "validate"

File: app-modules/events/database/factories/EventFactory.php [hunk 12]
```diff
@@ -32,4 +32,31 @@ public function definition(): array
...
+
+    public function past(): static
+    {
+        $startsAt = Date::now()->subDays(7);
+
+        return $this->state(fn (): array => [
+            'starts_at' => $startsAt,
+            'ends_at' => $startsAt->clone()->addHours(3),
+        ]);
+    }
```
Comment (danielhe4rt) [inline app-modules/events/database/factories/EventFactory.php:65]: Isso é ótimo, mas podemos dar uma melhorada no processo:

Hoje pra seedar um evento completo (com `EnrollmentPolicy`) são dois passos manuais:

```php
$event = Event::factory()->published()->create();
EnrollmentPolicy::create(['event_id' => $event->id, 'enrollment_method' => ..., ...]);
```

A ideia é adicionar states na `EventFactory` que representem tipos reais de evento da comunidade, cada um com um `afterCreating` que já cria a policy correta automaticamente. Algo assim:

```php
// Uso desejado
Event::factory()->asWorkshop()->create();
Event::factory()->asMeetup()->create();
Event::factory()->asConference()->create();
```

**Exemplo de implementação na `EventFactory`:**

```php
public function asWorkshop(): static
{
    return $this->state(fn (): array => [
        'event_type' => EventType::Workshop,
        'status' => EventStatus::Published,
    ])->afterCreating(function (Event $event): void {
        EnrollmentPolicy::factory()->create([
            'event_id' => $event->id,
            'enrollment_method' => EnrollmentMethod::RsvpCheckin,
            'check_in_method' => CheckInMethod::NumericCode,
            'capacity' => 50,
            'has_waitlist' => true,
            'attendance_requirement' => AttendanceRequirement::AllDays,
            'xp_on_confirmed' => 100,
            'xp_on_checked_in' => 200,
            'xp_on_attended' => 500,
        ]);
    });
}

public function asMeetup(): static
{
    return $this->state(fn (): array => [
        'event_type' => EventType::Meetup,
        'status' => EventStatus::Published,
    ])->afterCreating(function (Event $event): void {
        EnrollmentPolicy::factory()->create([
            'event_id' => $event->id,
            'enrollment_method' => EnrollmentMethod::Rsvp,
            'check_in_method' => CheckInMethod::Manual,
            'capacity' => null,
            'has_waitlist' => false,
            'attendance_requirement' => AttendanceRequirement::AllDays,
            'xp_on_confirmed' => 50,
            'xp_on_checked_in' => 100,
            'xp_on_attended' => 250,
        ]);
    });
}

public function asConference(): static
{
    return $this->state(fn (): array => [
        'event_type' => EventType::Conference,
        'status' => EventStatus::Published,
    ])->afterCreating(function (Event $event): void {
        EnrollmentPolicy::factory()->create([
            'event_id' => $event->id,
            'enrollment_method' => EnrollmentMethod::Application,
            'check_in_method' => CheckInMethod::QrCode,
            'capacity' => 200,
            'has_waitlist' => true,
            'attendance_requirement' => AttendanceRequirement::MinimumDays,
            'minimum_days' => 2,
            'cancellation_deadline_hours' => 48,
            'xp_on_confirmed' => 200,
            'xp_on_checked_in' => 300,
            'xp_on_attended' => 1000,
        ]);
    });
}
```

**Benefícios:**
- Seedar/testar fica one-liner: `Event::factory(10)->asMeetup()->create()`
- Testes de enrollment/check-in não precisam montar policy manualmente
- Os presets refletem a realidade da comunidade (workshops com código numérico, meetups abertos, conf com application + QR)
- Ainda dá pra sobrescrever campos via `->create([...])` normalmente

**Composição continua funcionando:**
```php
Event::factory()->asWorkshop()->upcoming()->create(['tenant_id' => 1]);
Event::factory(5)->asMeetup()->past()->create();
```

**Review by danielhe4rt — CHANGES_REQUESTED** [review]
Tamo quase lá! Primeiramente obrigado pela contribuição, só faltou umas pontas pra melhorar legibilidade e de resto tá tudo redondo e testado!

**Review by danielhe4rt — APPROVED** [review]
